# Deploy a BigQuery ML user churn propensity model to Vertex AI for online predictions

## Learning objectives

* Explore and preprocess a [Google Analytics 4](https://support.google.com/analytics/answer/7029846) data sample in [BigQuery](https://cloud.google.com/bigquery) for machine learning.  
* Train a [BigQuery ML (BQML)](https://cloud.google.com/bigquery-ml) [XGBoost](https://xgboost.readthedocs.io/en/latest/) classifier to predict user churn on a mobile gaming application.
* Tune a BQML XGBoost classifier using [BQML hyperparameter tuning features](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-create-boosted-tree).
* Evaluate the performance of a BQML XGBoost classifier.
* Explain your XGBoost model with [BQML Explainable AI](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-xai-overview) global feature attributions.
* Generate batch predictions with your BQML XGBoost model.
* Export a BQML XGBoost model to a [Google Cloud Storage](https://cloud.google.com/storage).
* Upload and deploy a BQML XGBoost model to a [Vertex AI Prediction](https://cloud.google.com/vertex-ai/docs/predictions/getting-predictions) Endpoint for online predictions.

## Introduction

In this lab, you will train, evaluate, explain, and generate batch and online predictions with a BigQuery ML (BQML) XGBoost model. You will use a Google Analytics 4 dataset from a real mobile application, Flood it! ([Android app](https://play.google.com/store/apps/details?id=com.labpixies.flood), [iOS app](https://itunes.apple.com/us/app/flood-it!/id476943146?mt=8)), to determine the likelihood of users returning to the application. You will generate batch predictions with your BigQuery ML model as well as export and deploy it to **Vertex AI** for online predictions.

[BigQuery ML](https://cloud.google.com/bigquery-ml/docs/introduction) lets you train and do batch inference with machine learning models in BigQuery using standard SQL queries faster by eliminating the need to move data with fewer lines of code. [Vertex AI](https://cloud.google.com/vertex-ai) is Google Cloud's complimentary next generation, unified platform for machine learning development. By developing and deploying BQML machine learning solutions on Vertex AI, you can leverage a scalable online prediction service and MLOps tools for model retraining and monitoring to significantly enhance your development productivity, the ability to scale your workflow and decision making with your data, and accelerate time to value.

![BQML Vertex AI](./images/vertex-bqml-lab-architecture-diagram.png "Vertex BQML Lab Architecture Diagram")

Note: this lab is inspired by and extends [Churn prediction for game developers using Google Analytics 4 (GA4) and BigQuery ML](https://cloud.google.com/blog/topics/developers-practitioners/churn-prediction-game-developers-using-google-analytics-4-ga4-and-bigquery-ml). See that blog post and accompanying tutorial for additional depth on this use case and BigQuery ML. In this lab, you will go one step further and focus on how Vertex AI extends BQML's capabilities through online prediction so you can incorporate both customer churn predictions into decision making UIs such as [Looker dashboards](https://looker.com/google-cloud) but also online predictions directly into customer applications to power targeted interventions such as targeted incentives.

### Use case: user churn propensity modeling in the mobile gaming industry

According to a [2019 study](https://gameanalytics.com/reports/mobile-gaming-industry-analysis-h1-2019) on 100K mobile games by the Mobile Gaming Industry Analysis, most mobile games only see a 25% retention rate for users after the first 24 hours, known and any game "below 30% retention generally needs improvement". For mobile game developers, improving user retention is critical to revenue stability and increasing profitability. In fact, [Bain & Company research](https://hbr.org/2014/10/the-value-of-keeping-the-right-customers) found that 5% growth in retention rate can result in a 25-95% increase in profits. With lower costs to retain existing customers, the business objective for game developers is clear: reduce churn and improve customer loyalty to drive long-term profitability.

Your task in this lab: use machine learning to predict user churn propensity after day 1, a crucial user onboarding window, and serve these online predictions to inform interventions such as targeted in-game rewards and notifications.

## Setup

In [1]:
!pip3 install google-cloud-aiplatform --user
!pip3 install pyarrow==11.0.0 --user
!pip3 install --upgrade google-cloud-bigquery --user
!pip3 install --upgrade google-cloud-bigquery-storage --user
!pip3 install --upgrade google-cloud-storage --user
!pip install db-dtypes

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.9/34.9 MB 44.0 MB/s eta 0:00:00:00:0100:01
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
apache-beam 2.46.0 requires grpcio!=1.48.0,<2,>=1.33.1, but you have grpcio 1.48.0 which is incompatible.
apache-beam 2.46.0 requires pyarrow<10.0.0,>=3.0.0, but you have pyarrow 11.0.0 which is incompatible.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 239.0/239.0 kB 5.3 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 199.8/199.8 kB 4.5 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.5/126.5 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.4/139.4 kB 16.5 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all t

**Restart the kernel and ignore the compatibility errors.**

### Define constants

In [2]:
# Retrieve and set PROJECT_ID and REGION environment variables.
PROJECT_ID = !(gcloud config get-value core/project)
PROJECT_ID = PROJECT_ID[0]

**Note:** Replace the <code>REGION</code> with the associated region mentioned in the qwiklabs resource panel.

In [3]:
BQ_LOCATION = 'US'
REGION = 'us-central1'

### Import libraries

In [4]:
from google.cloud import bigquery
from google.cloud import aiplatform as vertexai
import numpy as np
import pandas as pd

### Create a GCS bucket for artifact storage

Create a globally unique Google Cloud Storage bucket for artifact storage. You will use this bucket to export your BQML model later in the lab and upload it to Vertex AI.

In [5]:
GCS_BUCKET = f"{PROJECT_ID}-bqmlga4"

In [6]:
!gsutil mb -l $REGION gs://$GCS_BUCKET

Creating gs://qwiklabs-gcp-03-37bdd099dcdb-bqmlga4/...


### Create a BigQuery dataset

Next, create a BigQuery dataset from this notebook using the Python-based [`bq` command line utility](https://cloud.google.com/bigquery/docs/bq-command-line-tool). 

This dataset will group your feature views, model, and predictions table together. You can view it in the [BigQuery](https://pantheon.corp.google.com/bigquery) console.

In [7]:
BQ_DATASET = f"{PROJECT_ID}:bqmlga4"

In [8]:
!bq mk --location={BQ_LOCATION} --dataset {BQ_DATASET}

Dataset 'qwiklabs-gcp-03-37bdd099dcdb:bqmlga4' successfully created.


### Initialize the Vertex Python SDK client

Import the Vertex SDK for Python into your Python environment and initialize it.

In [9]:
vertexai.init(project=PROJECT_ID, location=REGION, staging_bucket=f"gs://{GCS_BUCKET}")

## Exploratory Data Analysis (EDA) in BigQuery

This lab uses a [public BigQuery dataset]() that contains raw event data from a real mobile gaming app called **Flood it!** ([Android app](https://play.google.com/store/apps/details?id=com.labpixies.flood), [iOS app](https://itunes.apple.com/us/app/flood-it!/id476943146?mt=8)).

The data schema originates from Google Analytics for Firebase but is the same schema as [Google Analytics 4](https://support.google.com/analytics/answer/9358801).

Take a look at a sample of the raw event dataset using the query below:

In [10]:
%%bigquery --project $PROJECT_ID

SELECT 
    *
FROM
  `firebase-public-project.analytics_153293282.events_*`
    
TABLESAMPLE SYSTEM (1 PERCENT)

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

event_date   event_timestamp       event_name  \
0       20180626  1530057986432000      screen_view   
1       20180626  1530057929539000    session_start   
2       20180626  1530057963525003  user_engagement   
3       20180626  1530057920265007  user_engagement   
4       20180626  1530057933838013  user_engagement   
...          ...               ...              ...   
49995   20180626  1530033510691013       post_score   
49996   20180626  1530033465543022       post_score   
49997   20180626  1530033406510002      screen_view   
49998   20180626  1530033610986002       post_score   
49999   20180626  1530033706741016       post_score   

                                            event_params  \
0      [{'key': 'firebase_screen_class', 'value': {'s...   
1      [{'key': 'firebase_screen_class', 'value': {'s...   
2      [{'key': 'firebase_screen_class', 'value': {'s...   
3      [{'key': 'firebase_screen_class', 'value': {'s...   
4      [{'key': 'firebase_screen_class', 'value': {'s...   
...                                                  ...   
49995  [{'key': 'firebase_screen_class', 'value': {'s...   
49996  [{'key': 'firebase_screen_class', 'value': {'s...   
49997  [{'key': 'firebase_screen_class', 'value': {'s...   
49998  [{'key': 'firebase_screen_class', 'value': {'s...   
49999  [{'key': 'firebase_screen_class', 'value': {'s...   

       event_previous_timestamp  event_value_in_usd  event_bundle_sequence_id  \
0              1529641186155000                 NaN                      <NA>   
1              1529640599918000                 NaN                      <NA>   
2              1529641198750003                 NaN                      <NA>   
3              1530057909525007                 NaN                      <NA>   
4              1530057912265013                 NaN                      <NA>   
...                         ...                 ...                       ...   
49995          1526918607216013                 NaN                      <NA>   
49996          1530033425691022                 NaN                      <NA>   
49997          1526918607768002                 NaN                      <NA>   
49998          1530033464543002                 NaN                      <NA>   
49999          1530033588986016                 NaN                      <NA>   

       event_server_timestamp_offset user_id  \
0                               <NA>    None   
1                               <NA>    None   
2                               <NA>    None   
3                               <NA>    None   
4                               <NA>    None   
...                              ...     ...   
49995                           <NA>    None   
49996                           <NA>    None   
49997                           <NA>    None   
49998                           <NA>    None   
49999                           <NA>    None   

                         user_pseudo_id  \
0      CF2898B41B7243671C36D5168D9D89B0   
1      CF2898B41B7243671C36D5168D9D89B0   
2      CF2898B41B7243671C36D5168D9D89B0   
3      CF2898B41B7243671C36D5168D9D89B0   
4      CF2898B41B7243671C36D5168D9D89B0   
...                                 ...   
49995  D2A4FF268B6047FDA0EC15B0365458F7   
49996  D2A4FF268B6047FDA0EC15B0365458F7   
49997  D2A4FF268B6047FDA0EC15B0365458F7   
49998  D2A4FF268B6047FDA0EC15B0365458F7   
49999  D2A4FF268B6047FDA0EC15B0365458F7   

                                         user_properties  \
0      [{'key': 'ad_frequency', 'value': {'string_val...   
1      [{'key': 'ad_frequency', 'value': {'string_val...   
2      [{'key': 'ad_frequency', 'value': {'string_val...   
3      [{'key': 'ad_frequency', 'value': {'string_val...   
4      [{'key': 'ad_frequency', 'value': {'string_val...   
...                                                  ...   
49995  [{'key': 'first_open_time', 'value': {'string_...   
49996  [{'key': 'first_open_time', 'value': {'string_...   
49997  [{'key': '

Note: in the cell above, Jupyterlab runs cells starting with `%%bigquery` as SQL queries. 

Google Analytics 4 uses an event based measurement model and each row in this dataset is an event. View the [complete schema](https://support.google.com/analytics/answer/7029846) and details about each column. As you can see above, certain columns are nested records and contain detailed information such as:

* app_info
* device
* ecommerce
* event_params
* geo
* traffic_source
* user_properties
* items*
* web_info*

This dataset contains 5.7M events from 15K+ users.

## Dataset preparation in BigQuery

Now that you have a better sense for the dataset you will be working with, you will walk through transforming raw event data into a dataset suitable for machine learning using SQL commands in BigQuery. Specifically, you will:

* Aggregate events so that each row represents a separate unique user ID.
* Define the **user churn label** feature to train your model to prediction (e.g. 1 = churned, 0 = returned).
* Create **user demographic** features.
* Create **user behavioral** features from aggregated application events.

### Defining churn for each user

There are many ways to define user churn, but for the purposes of this lab, you will predict 1-day churn as users who do not come back and use the app again after 24 hr of the user's first engagement. This is meant to capture churn after a user's "first impression" of the application or onboarding experience.

In other words, after 24 hr of a user's first engagement with the app:

* if the user shows no event data thereafter, the user is considered **churned**.
* if the user does have at least one event datapoint thereafter, then the user is considered **returned**.

You may also want to remove users who were unlikely to have ever returned anyway after spending just a few minutes with the app, which is sometimes referred to as "bouncing". For example, you will build your model on only on users who spent at least 10 minutes with the app (users who didn't bounce).

The query below defines a churned user with the following definition:

**Churned = "any user who spent at least 10 minutes on the app, but after 24 hour from when they first engaged with the app, never used the app again"**

You will use the raw event data, from their first touch (app installation) to their last touch, to identify churned and bounced users in the `user_churn` view query below:

In [11]:
%%bigquery --project $PROJECT_ID

CREATE OR REPLACE VIEW bqmlga4.user_churn AS (
  WITH firstlasttouch AS (
    SELECT
      user_pseudo_id,
      MIN(event_timestamp) AS user_first_engagement,
      MAX(event_timestamp) AS user_last_engagement
    FROM
      `firebase-public-project.analytics_153293282.events_*`
    WHERE event_name="user_engagement"
    GROUP BY
      user_pseudo_id

  )
  
SELECT
    user_pseudo_id,
    user_first_engagement,
    user_last_engagement,
    EXTRACT(MONTH from TIMESTAMP_MICROS(user_first_engagement)) as month,
    EXTRACT(DAYOFYEAR from TIMESTAMP_MICROS(user_first_engagement)) as julianday,
    EXTRACT(DAYOFWEEK from TIMESTAMP_MICROS(user_first_engagement)) as dayofweek,

    #add 24 hr to user's first touch
    (user_first_engagement + 86400000000) AS ts_24hr_after_first_engagement,
    
    #churned = 1 if last_touch within 24 hr of app installation, else 0
    IF (user_last_engagement < (user_first_engagement + 86400000000),
    1,
    0 ) AS churned,
    
    #bounced = 1 if last_touch within 10 min, else 0
    IF (user_last_engagement <= (user_first_engagement + 600000000),
    1,
    0 ) AS bounced,
  FROM
    firstlasttouch
  GROUP BY
    user_pseudo_id,
    user_first_engagement,
    user_last_engagement
    );

SELECT 
  * 
FROM 
  bqmlga4.user_churn 
LIMIT 100;

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

user_pseudo_id  user_first_engagement  \
0   9D535D543FD07B8247DEADB4669FADE6       1531154706915004   
1   5217AB1A454DAED6243E1C9818BE6A20       1529861523949001   
2   0A9DC17E5750D98D659892A5063E9F04       1531251327493000   
3   41CA0FA528C7227EA1E592DCC4733C40       1528914965271010   
4   7566596A1D6ACA781692A7A0B89B06EF       1528843166489001   
..                               ...                    ...   
95  9F056AEC0D7B47E9352A64B94B71DC74       1529312516854010   
96  FCC37EB1434C9A00F7D2D3FE06349C9A       1529354400099007   
97  51825D61E76B9BBC83F0411D14E6D1F2       1528813579200005   
98  AEC162E61CAF66AC209F108733FB4486       1529339057260003   
99  76FEA87933DC267A66B434A24E80AB7C       1529373627151003   

    user_last_engagement  month  julianday  dayofweek  \
0       1531764064336034      7        190          2   
1       1535405434229018      6        175          1   
2       1538578518951032      7        191          3   
3       1538495313957016      6        164          4   
4       1538356070861018      6        163          3   
..                   ...    ...        ...        ...   
95      1530490786112003      6        169          2   
96      1535095334209003      6        169          2   
97      1530038848567000      6        163          3   
98      1536606247295000      6        169          2   
99      1530494942212003      6        170          3   

    ts_24hr_after_first_engagement  churned  bounced  
0                 1531241106915004        0        0  
1                 1529947923949001        0        0  
2                 1531337727493000        0        0  
3                 1529001365271010        0        0  
4                 1528929566489001        0        0  
..                             ...      ...      ...  
95                1529398916854010        0        0  
96                1529440800099007        0        0  
97                1528899979200005        0        0  
98                1529425457260003        0        0  
99                1529460027151003        0        0  

[100 rows x 9 columns]

Review how many of the 15k users bounced and returned below:

In [12]:
%%bigquery --project $PROJECT_ID

SELECT
    bounced,
    churned, 
    COUNT(churned) as count_users
FROM
    bqmlga4.user_churn
GROUP BY 
  bounced,
  churned
ORDER BY 
  bounced

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

bounced  churned  count_users
0        0        0         6148
1        0        1         1883
2        1        1         5557

For the training data, you will only end up using data where bounced = 0. Based on the 15k users, you can see that 5,557 ( about 41%) users bounced within the first ten minutes of their first engagement with the app. Of the remaining 8,031 users, 1,883 users ( about 23%) churned after 24 hours which you can validate with the query below:

### Extract user demographic features

There is various user demographic information included in this dataset, including `app_info`, `device`, `ecommerce`, `event_params`, and `geo`. Demographic features can help the model predict whether users on certain devices or countries are more likely to churn.

Note that a user's demographics may occasionally change (e.g. moving countries). For simplicity, you will use the demographic information that Google Analytics 4 provides when the user first engaged with the app as indicated by MIN(event_timestamp) in the query below. This enables every unique user to be represented by a single row.

In [13]:
%%bigquery --project $PROJECT_ID

CREATE OR REPLACE VIEW bqmlga4.user_demographics AS (

  WITH first_values AS (
      SELECT
          user_pseudo_id,
          geo.country as country,
          device.operating_system as operating_system,
          device.language as language,
          ROW_NUMBER() OVER (PARTITION BY user_pseudo_id ORDER BY event_timestamp DESC) AS row_num
      FROM `firebase-public-project.analytics_153293282.events_*`
      WHERE event_name="user_engagement"
      )
  SELECT * EXCEPT (row_num)
  FROM first_values
  WHERE row_num = 1
  );

SELECT
  *
FROM
  bqmlga4.user_demographics
LIMIT 10

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

user_pseudo_id        country operating_system language
0  0043CF8841C819015ACD25710D333113       Pakistan          ANDROID    en-gb
1  00B6295EC351DBFE0E31015296871E35  United States          ANDROID    en-us
2  032918145D0FE92E31B0D5194DB1A4C3  United States          ANDROID    en-us
3  03AD106A71792A3FCC1D5DFA9796E404  United States              IOS    en-us
4  0492FB77E1C23B1993CC0B871AACC8C2  United States          ANDROID    en-us
5  0B2A8B21C42C0B9899D864F91BF109E6         Canada             None    en-ca
6  0C823A0351DED69746589F995D0C682C         Russia          ANDROID    ru-ru
7  1170309EA1D2309A158B8F6346AECE88      Australia          ANDROID    en-au
8  1458DD2FE0BA99EE08FE02D37DFA8058  United States          ANDROID    en-us
9  1647132022710A68A8C58B24BD0C5965  United States              IOS    en-us

### Aggregate user behavioral features

Behavioral data in the raw event data spans across multiple events -- and thus rows -- per user. The goal of this section is to aggregate and extract behavioral data for each user, resulting in one row of behavioral data per unique user.



As a first step, you can explore all the unique events that exist in this dataset, based on event_name:

In [14]:
%%bigquery --project $PROJECT_ID

SELECT
  event_name,
  COUNT(event_name) as event_count
FROM
    `firebase-public-project.analytics_153293282.events_*`
GROUP BY 
  event_name
ORDER BY
   event_count DESC

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

event_name  event_count
0                screen_view      2247623
1            user_engagement      1358958
2      level_start_quickplay       523430
3        level_end_quickplay       349729
4                 post_score       242051
5   level_complete_quickplay       191088
6       level_fail_quickplay       137035
7      level_reset_quickplay       122278
8             select_content       105139
9                level_start        74417
10             session_start        74353
11                 level_end        54582
12               level_retry        43345
13                  level_up        35666
14            level_complete        33986
15     level_retry_quickplay        29939
16                level_fail        17353
17    spend_virtual_currency         9363
18           use_extra_steps         8945
19               level_reset         8535
20         firebase_campaign         5411
21             app_exception         5156
22                first_open         4322
23                 os_update         3290
24       no_more_extra_steps         2792
25                app_remove         2639
26                app_update         2473
27                 ad_reward         1912
28        completed_5_levels         1867
29                     error         1714
30        challenge_a_friend          524
31           in_app_purchase           27
32     dynamic_link_app_open           19
33            app_clear_data           17
34   dynamic_link_first_open           16
35        challenge_accepted            5
36   notification_foreground            1

For this lab, to predict whether a user will churn or return, you can start by counting the number of times a user engages in the following event types:

* user_engagement
* level_start_quickplay
* level_end_quickplay
* level_complete_quickplay
* level_reset_quickplay
* post_score
* spend_virtual_currency
* ad_reward
* challenge_a_friend
* completed_5_levels
* use_extra_steps

In the SQL query below, you will aggregate the behavioral data by calculating the total number of times when each of the above event_names occurred in the data set per user.

In [15]:
%%bigquery --project $PROJECT_ID

CREATE OR REPLACE VIEW bqmlga4.user_behavior AS (
WITH
  events_first24hr AS (
    # Select user data only from first 24 hr of using the app.
    SELECT
      e.*
    FROM
      `firebase-public-project.analytics_153293282.events_*` e
    JOIN
      bqmlga4.user_churn c
    ON
      e.user_pseudo_id = c.user_pseudo_id
    WHERE
      e.event_timestamp <= c.ts_24hr_after_first_engagement
    )
SELECT
  user_pseudo_id,
  SUM(IF(event_name = 'user_engagement', 1, 0)) AS cnt_user_engagement,
  SUM(IF(event_name = 'level_start_quickplay', 1, 0)) AS cnt_level_start_quickplay,
  SUM(IF(event_name = 'level_end_quickplay', 1, 0)) AS cnt_level_end_quickplay,
  SUM(IF(event_name = 'level_complete_quickplay', 1, 0)) AS cnt_level_complete_quickplay,
  SUM(IF(event_name = 'level_reset_quickplay', 1, 0)) AS cnt_level_reset_quickplay,
  SUM(IF(event_name = 'post_score', 1, 0)) AS cnt_post_score,
  SUM(IF(event_name = 'spend_virtual_currency', 1, 0)) AS cnt_spend_virtual_currency,
  SUM(IF(event_name = 'ad_reward', 1, 0)) AS cnt_ad_reward,
  SUM(IF(event_name = 'challenge_a_friend', 1, 0)) AS cnt_challenge_a_friend,
  SUM(IF(event_name = 'completed_5_levels', 1, 0)) AS cnt_completed_5_levels,
  SUM(IF(event_name = 'use_extra_steps', 1, 0)) AS cnt_use_extra_steps,
FROM
  events_first24hr
GROUP BY
  user_pseudo_id
  );

SELECT
  *
FROM
  bqmlga4.user_behavior
LIMIT 10

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

user_pseudo_id  cnt_user_engagement  \
0  9B6D9BA83A59A7BBB805D8656DF47BCF                   21   
1  DA426D16B0EAB27DE1AC9D6E3418F400                  208   
2  6A50ADC841633E84CD6BA0C528FE007C                  180   
3  F8B95AE4A257A86F7E55E7C80935A4E2                   16   
4  E80C958E3C33898989A30B8F5A4514F0                   40   
5  72F61883EF67515CBBFEFBBDFF93210A                   23   
6  A98E72E7110470331C8EDB822E9A1D8D                   22   
7  2CD70A7B8E8442DBF6BB7E5EFEEFA0B1                   27   
8  9931BF1870F9193FAD87597955E5CD4C                    2   
9  B6A8AEA27748774830324A996FFE2687                   11   

   cnt_level_start_quickplay  cnt_level_end_quickplay  \
0                          0                        0   
1                          6                        4   
2                         42                       41   
3                          0                        0   
4                          3                        1   
5                          0                        0   
6                          2                        1   
7                          0                        0   
8                          1                        0   
9                          6                        1   

   cnt_level_complete_quickplay  cnt_level_reset_quickplay  cnt_post_score  \
0                             0                          0               4   
1                             0                          0              30   
2                             4                          0               4   
3                             0                          0               2   
4                             0                          0               5   
5                             0                          0              10   
6                             1                          0               5   
7                             0                          0               6   
8                             0                          0               0   
9                             0                          4               0   

   cnt_spend_virtual_currency  cnt_ad_reward  cnt_challenge_a_friend  \
0                           1              0                       0   
1                          11              2                       0   
2                          12              1                       0   
3                           1              0                       0   
4                           0              0                       0   
5                           0              0                       0   
6                           0              0                       0   
7                           0              0                       0   
8                           0              0                       0   
9                           0              0                       0   

   cnt_completed_5_levels  cnt_use_extra_steps  
0                       0                    1  
1                       1                   11  
2                       0                   12  
3                       0                    1  
4                       1                    0  
5                       1                    0  
6                       0                    0  
7                       1                    0  
8                       0                    0  
9                       0                    0

### Prepare your train/eval/test datasets for machine learning

In this section, you can now combine these three intermediary views (`user_churn`, `user_demographics`, and `user_behavior`) into the final training data view called `ml_features`. Here you can also specify bounced = 0, in order to limit the training data only to users who did not "bounce" within the first 10 minutes of using the app.

Note in the query below that a manual `data_split` column is created in your BigQuery ML table using [BigQuery's hashing functions](https://towardsdatascience.com/ml-design-pattern-5-repeatable-sampling-c0ccb2889f39) for repeatable sampling. It specifies a 80% train | 10% eval | 20% test split to evaluate your model's performance and generalization.

In [16]:
%%bigquery --project $PROJECT_ID

CREATE OR REPLACE VIEW bqmlga4.ml_features AS (
    
  SELECT
    dem.user_pseudo_id,
    IFNULL(dem.country, "Unknown") AS country,
    IFNULL(dem.operating_system, "Unknown") AS operating_system,
    IFNULL(REPLACE(dem.language, "-", "X"), "Unknown") AS language,
    IFNULL(beh.cnt_user_engagement, 0) AS cnt_user_engagement,
    IFNULL(beh.cnt_level_start_quickplay, 0) AS cnt_level_start_quickplay,
    IFNULL(beh.cnt_level_end_quickplay, 0) AS cnt_level_end_quickplay,
    IFNULL(beh.cnt_level_complete_quickplay, 0) AS cnt_level_complete_quickplay,
    IFNULL(beh.cnt_level_reset_quickplay, 0) AS cnt_level_reset_quickplay,
    IFNULL(beh.cnt_post_score, 0) AS cnt_post_score,
    IFNULL(beh.cnt_spend_virtual_currency, 0) AS cnt_spend_virtual_currency,
    IFNULL(beh.cnt_ad_reward, 0) AS cnt_ad_reward,
    IFNULL(beh.cnt_challenge_a_friend, 0) AS cnt_challenge_a_friend,
    IFNULL(beh.cnt_completed_5_levels, 0) AS cnt_completed_5_levels,
    IFNULL(beh.cnt_use_extra_steps, 0) AS cnt_use_extra_steps,
    chu.user_first_engagement,
    chu.month,
    chu.julianday,
    chu.dayofweek,
    chu.churned,
    # https://towardsdatascience.com/ml-design-pattern-5-repeatable-sampling-c0ccb2889f39
    # BQML Hyperparameter tuning requires STRING 3 partition data_split column.
    # 80% 'TRAIN' | 10%'EVAL' | 10% 'TEST'    
    CASE
      WHEN ABS(MOD(FARM_FINGERPRINT(dem.user_pseudo_id), 10)) <= 7
        THEN 'TRAIN'
      WHEN ABS(MOD(FARM_FINGERPRINT(dem.user_pseudo_id), 10)) = 8
        THEN 'EVAL'
      WHEN ABS(MOD(FARM_FINGERPRINT(dem.user_pseudo_id), 10)) = 9
        THEN 'TEST'    
          ELSE '' END AS data_split
  FROM
    bqmlga4.user_churn chu
  LEFT OUTER JOIN
    bqmlga4.user_demographics dem
  ON 
    chu.user_pseudo_id = dem.user_pseudo_id
  LEFT OUTER JOIN 
    bqmlga4.user_behavior beh
  ON
    chu.user_pseudo_id = beh.user_pseudo_id
  WHERE chu.bounced = 0
  );

SELECT
  *
FROM
  bqmlga4.ml_features
LIMIT 10

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

user_pseudo_id         country operating_system language  \
0  5217AB1A454DAED6243E1C9818BE6A20   United States          ANDROID    enXus   
1  0155220BAB309646CC7AC980AB75A212   United States          ANDROID    enXus   
2  41CA0FA528C7227EA1E592DCC4733C40   United States          ANDROID    enXus   
3  7566596A1D6ACA781692A7A0B89B06EF   United States          ANDROID    enXus   
4  36CD9DA75188D11BA0065964FF116EC0           India          ANDROID    enXin   
5  3F6BC08D796DEE0902E221D5ADC88FC5   United States          ANDROID    enXus   
6  8320FC247A1967E7C0CDFC1502241310   United States          ANDROID    enXus   
7  4789C778386485B99F4077B97DFE34E3   United States          ANDROID    enXus   
8  E512079B53179DF9A608CF4ADE47DE9D       Australia          ANDROID    enXau   
9  0FE7AA892150DE7C58D1B2988580346D  United Kingdom          ANDROID    enXgb   

   cnt_user_engagement  cnt_level_start_quickplay  cnt_level_end_quickplay  \
0                   89                          0                        0   
1                   46                         17                       16   
2                   25                         11                        9   
3                    8                          1                        1   
4                  110                          0                        0   
5                   39                         18                       16   
6                   38                          0                        0   
7                  192                        281                       46   
8                    5                          2                        1   
9                    9                          0                        0   

   cnt_level_complete_quickplay  cnt_level_reset_quickplay  cnt_post_score  \
0                             0                          0              27   
1                            11                          0              11   
2                             7                          1               7   
3                             0                          0               0   
4                             0                          0              26   
5                            14                          2              14   
6                             0                          0               4   
7                            33                        234              33   
8                             0                          0               1   
9                             0                          0               1   

   ...  cnt_ad_reward  cnt_challenge_a_friend  cnt_completed_5_levels  \
0  ...              0                       0                       1   
1  ...              0                       0                       0   
2  ...              0                       0                       0   
3  ...              0                       0                       0   
4  ...              0                       0                       1   
5  ...              0                       0                       0   
6  ...              0                       0                       0   
7  ...              0                       0                       0   
8  ...              0                       0                       0   
9  ...              1                       0                       0   

   cnt_use_extra_steps  user_first_engagement  month  julianday  dayofweek  \
0                    1       1529861523949001      6        175          1   
1                    0       1530818377077004      7        186          5   
2                    0       1528914965271010      6        164          4   
3                    0       1528843166489001      6        163          3   
4                    6       1530045756932001      6        177          3   
5                    0       1529645265072001      6        173          6   
6                    0       1530543398463008      7        1

### Validate feature splits

Run the query below to validate the number of examples in each data partition for the 80% train |10% eval |10% test split.

In [17]:
%%bigquery --project $PROJECT_ID

SELECT
  data_split,
  COUNT(*) AS n_examples
FROM bqmlga4.ml_features
GROUP BY data_split

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

data_split  n_examples
0      TRAIN        6386
1       EVAL         846
2       TEST         799

## Train and tune a BQML XGBoost propensity model to predict customer churn

The following code trains and tunes the hyperparameters for an XGBoost model. TO provide a minimal demonstration of BQML hyperparameter tuning in this lab, this model will take about 18 min to train and tune with its restricted search space and low number of trials. In practice, you would generally want at [least 10 trials per hyperparameter](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-hyperparameter-tuning#how_many_trials_do_i_need_to_tune_a_model) to achieve improved results.

For more information on the default hyperparameters used, you can read the documentation:
[CREATE MODEL statement for Boosted Tree models using XGBoost](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-create-boosted-tree)

|Model   | BQML model_type | Advantages | Disadvantages| 
|:-------|:----------:|:----------:|-------------:|
|XGBoost |     BOOSTED_TREE_CLASSIFIER [(documentation)](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-create-boosted-tree)       |   High model performance with feature importances and explainability | Slower to train than BQML LOGISTIC_REG |

Note: When you run the CREATE MODEL statement, BigQuery ML can automatically split your data into training and test so you can immediately evaluate your model's performance after training. This is a great option for fast model prototyping. In this lab, however, you split your data manually above using hashing for reproducible data splits that can be used comparing model evaluations across different runs.

In [18]:
MODEL_NAME="churn_xgb"

In [40]:
%%bigquery --project $PROJECT_ID

CREATE OR REPLACE MODEL bqmlga4.churn_xgb

OPTIONS(
  MODEL_TYPE="BOOSTED_TREE_CLASSIFIER",
  # Declare label column.
  INPUT_LABEL_COLS=["churned"],
  # Specify custom data splitting using the `data_split` column.
  DATA_SPLIT_METHOD="CUSTOM",
  DATA_SPLIT_COL="data_split",
  # Enable Vertex Explainable AI aggregated feature attributions.
  ENABLE_GLOBAL_EXPLAIN=false,
  # Hyperparameter tuning arguments.
  num_trials=1,
  max_parallel_trials=5,
  HPARAM_TUNING_OBJECTIVES=["roc_auc"],
  EARLY_STOP=True,
  # Hyperpameter search space.
  LEARN_RATE=HPARAM_RANGE(0.01, 0.1),
  MAX_TREE_DEPTH=HPARAM_CANDIDATES([1,2])
) AS

SELECT
  * EXCEPT(user_pseudo_id)
FROM
  bqmlga4.ml_features
    LIMIT 50

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

## Evaluate BQML XGBoost model performance

Once training is finished, you can run [ML.EVALUATE](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-evaluate) to return model evaluation metrics. By default, all model trials will be returned so the below query just returns the model performance for optimal first trial.

In [41]:
%%bigquery --project $PROJECT_ID

SELECT
  *
FROM
  ML.EVALUATE(MODEL bqmlga4.churn_xgb)
WHERE trial_id=1;

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

trial_id  precision    recall  accuracy  f1_score  log_loss   roc_auc
0         1   0.604167  0.148718  0.768461  0.238683  0.480936  0.767545

ML.EVALUATE generates the [precision, recall](https://developers.google.com/machine-learning/crash-course/classification/precision-and-recall), [accuracy](https://developers.google.com/machine-learning/crash-course/classification/accuracy), [log_loss](https://en.wikipedia.org/wiki/Loss_functions_for_classification#Logistic_loss), [f1_score](https://en.wikipedia.org/wiki/F-score) and [roc_auc](https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc) using the default classification threshold of 0.5, which can be modified by using the optional `THRESHOLD` parameter.

Next, use the [ML.CONFUSION_MATRIX](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-confusion) function to return a confusion matrix for the input classification model and input data.

For more information on confusion matrices, you can read through a detailed explanation [here](https://developers.google.com/machine-learning/crash-course/classification/true-false-positive-negative).

You can also plot the AUC-ROC curve by using [ML.ROC_CURVE](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-roc) to return the metrics for different threshold values for the model.

## Inspect global feature attributions

To provide further context to your model performance, you can use the [ML.GLOBAL_EXPLAIN](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-global-explain#get_global_feature_importance_for_each_class_of_a_boosted_tree_classifier_model) function which leverages Vertex Explainable AI as a back-end. [Vertex Explainable AI](https://cloud.google.com/vertex-ai/docs/explainable-ai) helps you understand your model's outputs for classification and regression tasks. Specifically, Vertex AI tells you how much each feature in the data contributed to your model's predicted result. You can then use this information to verify that the model is behaving as expected, identify and mitigate biases in your models, and get ideas for ways to improve your model and your training data.

## Generate batch predictions

You can generate batch predictions for your BQML XGBoost model using [ML.PREDICT](https://cloud.google.com/bigquery-ml/docs/reference/standard-sql/bigqueryml-syntax-predict).

In [42]:
%%bigquery --project $PROJECT_ID

SELECT
  *
FROM
  ML.PREDICT(MODEL bqmlga4.churn_xgb,
  (SELECT * FROM bqmlga4.ml_features WHERE data_split = "TEST"))

Query is running:   0%|          |

/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2309: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2323: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)
/opt/conda/lib/python3.10/site-packages/google/cloud/bigquery/table.py:2337: UserWarning: Unable to represent RANGE schema as struct using pandas ArrowDtype. Using `object` instead. To use ArrowDtype, use pandas >= 1.5 and pyarrow >= 10.0.1.
  warnings.warn(_RANGE_PYARROW_WARNING)


Downloading:   0%|          |

trial_id  predicted_churned  \
0           1                  0   
1           1                  0   
2           1                  0   
3           1                  0   
4           1                  0   
..        ...                ...   
794         1                  0   
795         1                  0   
796         1                  0   
797         1                  0   
798         1                  1   

                               predicted_churned_probs  \
0    [{'label': 1, 'prob': 0.11438297480344772}, {'...   
1    [{'label': 1, 'prob': 0.2760365903377533}, {'l...   
2    [{'label': 1, 'prob': 0.45523977279663086}, {'...   
3    [{'label': 1, 'prob': 0.3964616060256958}, {'l...   
4    [{'label': 1, 'prob': 0.11139767616987228}, {'...   
..                                                 ...   
794  [{'label': 1, 'prob': 0.2531121075153351}, {'l...   
795  [{'label': 1, 'prob': 0.4032173454761505}, {'l...   
796  [{'label': 1, 'prob': 0.2862566113471985}, {'l...   
797  [{'label': 1, 'prob': 0.29275381565093994}, {'...   
798  [{'label': 1, 'prob': 0.7310328483581543}, {'l...   

                       user_pseudo_id         country operating_system  \
0    690DA199396D35B845005CCD13ED8D67         Germany          ANDROID   
1    75C0C246DE2188E02D984C2E71B7B4C0   United States              IOS   
2    9F8D4A18FC571082BD1D7AA5597756AD           India          ANDROID   
3    C28231099C78071ADF8D0BAEA7B9C1F2  United Kingdom              IOS   
4    74456352647C13028F89D1EEFFB905F4   United States              IOS   
..                                ...             ...              ...   
794  0E75AB4A1A6C0A0D782DC4D7672F54BF   United States              IOS   
795  4E246E12AC481377A56103BFDA4AEEEF   United States              IOS   
796  7DFC677A7E3FB2E644F40FF5EFF901D3    Saudi Arabia          ANDROID   
797  3B639DBE0003532E333453BDF363FEB0   United States              IOS   
798  9E77E417DC513E9A797BDBBED2388BA3   United States              IOS   

    language  cnt_user_engagement  cnt_level_start_quickplay  \
0      deXde                    3                          1   
1      enXus                   17                          5   
2      enXus                   49                         16   
3      enXgb                   43                         26   
4     esX419                   18                          5   
..       ...                  ...                        ...   
794    enXus                    8                          2   
795    enXus                   10                          0   
796    arXil                    6                          0   
797    enXus                   23                          1   
798    enXus                    7                          3   

     cnt_level_end_quickplay  ...  cnt_ad_reward  cnt_challenge_a_friend  \
0                          1  ...              0                       0   
1                          4  ...              0                       0   
2                          0  ...              0                       0   
3                         11  ...              0                       0   
4                          2  ...              0                       0   
..                       ...  ...            ...                     ...   
794                        1  ...              0                       0   
795                        0  ...              0                       0   
796                        0  ...              0                       0   
797                        0  ...              0                       0   
798                        0  ...              0                       0   

     cnt_completed_5_levels  cnt_use_extra_steps  user_first_engagement  \
0                         0                    0       1529315116058004   
1                         0                    0       1531698441624000   
2                         0                    1       1535883982068001   
3   

The following query returns the probability that the user will return after 24 hrs. The higher the probability and closer it is to 1, the more likely the user is predicted to churn, and the closer it is to 0, the more likely the user is predicted to return.

In [43]:
%%bigquery --project $PROJECT_ID

CREATE OR REPLACE TABLE bqmlga4.churn_predictions AS (
SELECT
  user_pseudo_id,
  churned,
  predicted_churned,
  predicted_churned_probs[OFFSET(0)].prob as probability_churned
FROM
  ML.PREDICT(MODEL bqmlga4.churn_xgb,
  (SELECT * FROM bqmlga4.ml_features))
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []